keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1457 @@
+3. Always using {} for structs and maps (no more obscure errors about mapping values).
+4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).
+
+Taken together,
+these mean that KYAML is not whitespace-sensitive,
+which makes templating and other text-patching much easier.
+YAML calls this “flow style”,
+as opposed to the conventional “block style”.
+
+This might sound a lot like JSON,
+because it is!
+Remember that JSON is valid YAML.
+KYAML is friendlier than JSON for humans
+to read,
+to write,
+and to commit to source control.
+Unlike JSON, KYAML:
+1. Allows comments (when authored by users).
+2. Allows trailing commas on the last item in a list or struct/map.
+3. Does not require quoted keys.
+
+KYAML also includes a “header” (still valid YAML)
+which helps to disambiguate a KYAML document from an ill-formed JSON document
+(see: [kubernetes/kubernetes#77946](https://github.com/kubernetes/kubernetes/issues/78946)).
+
+KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.
+
+Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.
+
+To make this change, we will:
+1. Create new code (see details below) which serializes any object into KYAML.
+2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
+3. Create tooling which reformats YAML or JSON as KYAML (see details below).
+4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
+5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
+6. Make a lot of noise about it, and try to get more people to use it
+
+The rationale for changing everything everywhere is simple.
+Many users start with examples,
+which they copy and modify.
+By changing the examples we seed the change in the ecosystem.
+One success metric might be
+that KYAML becomes the preferred format in the most common helm charts.
+
+<<[UNRESOLVED to gate or not to gate]>>
+Should we require an env var to enable kyaml in the first release?  Are there
+sig-cli conventions here that I don't know?
+<</UNRESOLVED>>
+
+### Notes/Constraints/Caveats
+
+**Important note:**
+This is EXCLUSIVELY about client-side formatting.
+This KEP *does not* propose any changes to apiserver
+or other components besides kubectl.
+
+Syntax has always been a magnet for arguments in our industry.
+We expect this to be no exception.
+Reasonable people will disagree with our choices,
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+Givent that KYAML is YAML it should be possible to revise
+without breaking anyone.
```

Comment:
Any byte-accurate tests consuming kyaml output will break (we saw this happen on yamlv2 to yamlv3 indent / wrapping changes). If kyaml output is not guaranteed to be byte-stable, we need to clearly / loudly state this, and ideally, do something in the implementation so someone depending on it being byte-stable notices (e.g. make some aspect of the output just *slightly* non-deterministic or something)

Reply:
Thoughts on how to prevent Hyrum's law?  I don't think we want to (or can, really) randomize field ordering, for example.  Is it sufficient to say "this is a bad idea" ?  Do we need to version KYAML?

Reply:
Added words.

Reply:
Let's not make the output non-deterministic. The "dump diff between two objects" usage would then show semantically irrelevant differences.

Reply:
I think it's unreasonable to depend on a local formatter being byte-stable between minor releases.
If the server were serving it, maybe, but that functionality is explicitly not included in this KEP.

Reply:
My use case is for invoking the formatter twice within the same process.

I think we all agree that no stability should be expected between releases, it's just a question of how to make users aware of that.